compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/syntax/tree/NodeFactory.java

```diff
@@ -0,0 +1,120 @@
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+package io.ballerinalang.compiler.syntax.tree;
+
+import io.ballerinalang.compiler.internal.parser.tree.STNode;
+import io.ballerinalang.compiler.internal.parser.tree.STNodeFactory;
+import io.ballerinalang.compiler.internal.parser.tree.STToken;
+import io.ballerinalang.compiler.internal.parser.tree.SyntaxKind;
+
+import java.util.ArrayList;
+import java.util.Objects;
+
+/**
+ * A factory for creating nodes in the syntax tree.
+ *
+ * @since 1.3.0
+ */
+public class NodeFactory {
+
+    private NodeFactory() {
+    }
+
+    public static ModulePart createModulePart(NodeList<ImportDeclaration> imports,
+                                              NodeList<ModuleMemberDeclaration> members,
+                                              Token eofToken) {
+        Objects.requireNonNull(imports, "imports must not be null");
+        Objects.requireNonNull(members, "members must not be null");
+        Objects.requireNonNull(eofToken, "eofToken must not be null");
+        STNode stModulePart = STNodeFactory.createModulePart(imports.underlyingListNode().internalNode(),
+                members.underlyingListNode().internalNode(), eofToken.internalNode());
+        return stModulePart.createUnlinkedFacade();
+    }
+
+    public static FunctionDefinitionNode createFunctionDefinitionNode(Token visibilityQualifier,
+                                                                      Token functionKeyword,
+                                                                      Identifier functionName,
+                                                                      Token openParenToken,
+                                                                      NodeList<Parameter> parameters,
+                                                                      Token closeParenToken,
+                                                                      Node returnTypeDesc,
+                                                                      BlockStatement functionBody) {
+        Objects.requireNonNull(visibilityQualifier, "visibilityQualifier must not be null");
+        Objects.requireNonNull(functionKeyword, "functionKeyword must not be null");
+        Objects.requireNonNull(functionName, "functionName must not be null");
+        Objects.requireNonNull(openParenToken, "openParenToken must not be null");
+        Objects.requireNonNull(parameters, "parameters must not be null");
+        Objects.requireNonNull(closeParenToken, "closeParenToken must not be null");
+        Objects.requireNonNull(returnTypeDesc, "returnTypeDesc must not be null");
+        Objects.requireNonNull(functionBody, "functionBody must not be null");
```

Comment:
Can we delegate this to a common method? something like: `validateNodes(Node... nodes)` that loops through and validates

Reply:
Can do that Supun? I’ve done that in each “modify” methods in nodes. 

But, if we do that, we won’t be able to provide precise error messages. That is why I wanted to do a null check for each parameter.

Reply:
On a different thought, I think we can do what you suggested if we store field names such as "visibilityQualifier", "functionKeyword", "functionName" etc. As we discussed during the call, Syntax Tree to diagram generator needs field names.